### ISSUE TITLE
Exceptions thrown with `error.throwWithMessage` can not be catched

### ISSUE THREAD

**Issue by flbulgarelli** [uqbar-project/wollok#963]
See the following REPL session and compare how `throw` and `error.throwWithMessage` behave differently: 

```
Wollok interactive console (type "quit" to quit):
>>> error
error[]
>>> error.throwWithMessage("foo")
wollok.lang.Exception: foo
	at wollok.lib.error.throwWithMessage(aMessage) [/lib.wlk:255]
	at  [/__synthetic0.wpgm:3]
	at  [/lib.wlk:249]
	at  [/wollokREPL.wlk:1]

>>> try { error.throwWithMessage("foo") } catch e : Exception { } 
wollok.lang.Exception: foo
	at wollok.lib.error.throwWithMessage(aMessage) [/lib.wlk:255]
	at  [/__synthetic0.wpgm:3]
	at  [/__synthetic0.wpgm:3]
	at  [/lib.wlk:249]
	at  [/wollokREPL.wlk:1]

>>> try { throw new Exception("foo") } catch e : Exception { } 
null
>>> try { throw new Exception("foo") } catch e : wollok.lang.Exception { } 
null
>>> try { error.throwWithMessage("foo") } catch e : wollok.lang.Exception { } 
wollok.lang.Exception: foo
	at wollok.lib.error.throwWithMessage(aMessage) [/lib.wlk:255]
	at  [/__synthetic0.wpgm:3]
	at  [/__synthetic0.wpgm:3]
	at  [/__synthetic0.wpgm:3]
	at  [/__synthetic0.wpgm:3]
	at  [/__synthetic0.wpgm:3]
	at  [/lib.wlk:249]
	at  [/wollokREPL.wlk:1]

>>> 
```

I think that this has sometime worked, but I am not 100% confident about that.

Is this easy to fix? Is there any workaround to this bug? It is somewhat blocking for me since it breaks the mumuki console and prevents to run many exercises.

**Comment by javierfernandes** [comment]
There are at least 2 different issues here

1) Repl shouldn't show output for empty catchs

```scala
>>> try { throw new Exception("foo") } catch e : Exception { } 
null
```

The piece of code doesn't produce any value, so ideally it shouldn't log anything.

But still the execution semantic of the interpreter is right here. You are catching the same exception you are throwing. It must not pop to the repl.

2) Exception thrown by the "error" object is not being catched !

```try { error.throwWithMessage("foo") } catch e : wollok.lang.Exception { }```

This should be catched and it shouldn't pop the exception further.
This issues is more important than the first one.
Not really sure why this is happening. Maybe the exceptions is being wrapped.. not sure.

**Comment by flbulgarelli** [comment]
Yes, the issue I am reporting here is the latter - the first is an issue too, but is is non-blocking, and I think it happens with `if`s too. 

Just a hint: the second issue doesn't only happen in the REPL, but also though the wollok server, within a `progam` block.

**Comment by fdodino** [comment]
try { error.throwWithMessage("foo") } catch e : Exception {  } 

While same line in Wollok program works fine, in REPL it throws an exception.
In WollokInterpreter, rootObject are different

* rootObject: [org.uqbar.project.wollok.wollokDsl.impl.ImportImpl@4f3faa70 (importedNamespace: wollokREPL.*) that's for REPL
* rootObject: [org.uqbar.project.wollok.wollokDsl.impl.WProgramImpl@6a9d5dff (name: abc)] : that works for a Program.

I'll continue my research.

**Comment by fdodino** [comment]
Ok, problem is that in WollokInterpreterEvaluator

```scala
	def dispatch evaluate(WTry t) {
		try
			t.expression.eval
		catch (WollokProgramExceptionWrapper e) {
			val cach = t.catchBlocks.findFirst[	it.matches(e.wollokException) 	]
			if (cach != null) {
				cach.evaluate(e)
			} else
				throw e
		} finally
			t.alwaysExpression?.eval
	}
```

It calls "matches" extension method:

```scala
	def boolean matches(WCatch cach, WollokObject it) {
		cach.exceptionType == null || isKindOf(cach.exceptionType)
	}
```

and isKindOf is 

```scala
def isKindOf(WMethodContainer c) { behavior.isKindOf(c)	}
```

But... while in program they are the same class:
behavior: org.uqbar.project.wollok.wollokDsl.impl.WClassImpl@26bab2f1 (name: Exception)
c: org.uqbar.project.wollok.wollokDsl.impl.WClassImpl@6d4c273c (name: Exception)

In REPL they are different classes:
behavior: org.uqbar.project.wollok.wollokDsl.impl.WClassImpl@26bab2f1 (name: Exception)
c: org.uqbar.project.wollok.wollokDsl.impl.WClassImpl@6d4c273c (name: Exception)

So... this method return false, and an exception is thrown:

```scala
	def static dispatch isKindOf(WClass c1, WClass c2) {
		WollokModelExtensions.isSuperTypeOf(c2, c1)
	}
```

WollokModelExtension
```scala
	def static boolean isSuperTypeOf(WClass a, WClass b) {
		a == b || (b.parent != null && a.isSuperTypeOf(b.parent))
	}
```

But I don't get why... I feel like an economic analyst.

**Comment by javierfernandes** [comment]
Nice analysis !!

Basically, WClasses equality is currently based on identity. And somehow on the REPL the base classes of wollok (SDK) are being loaded twice (or more times?).
So when it tries to compare two classes they don't match.

I would be good to understand the background fact that is loading SDK many times.
But, another options is to change class equality to something like a value object unrelated to instances.

```xtend
 def static boolean isSuperTypeOf(WClass a, WClass b) {
        a == b || (b.parent != null && a.isSuperTypeOf(b.parent))
    }
```
Could be something like

```xtend
 def static boolean isSuperTypeOf(WClass a, WClass b) {
        a.fqn == b.fqn || (b.parent != null && a.isSuperTypeOf(b.parent))
    }
```

Notice the "fqn". I'm not sure but I believe that we have one extension method with that name somewhere. It gives you the full element name with package and all.

**Comment by fdodino** [comment]
:+1: 

Yes, I was thinking that as a workaround, but now I'm more confident because you suggested so.
:smile:

Nevertheless I'll try to see what is going on here a little bit more.

**Comment by fdodino** [comment]
fqn is an extension method valid

**Comment by npasserini** [comment]
:)

**Comment by fdodino** [comment]
Well, I fighted and fighted, but I can't figure out where does Wollok create a second instance of Exception class. It happens only with error wko, throwing an error works, so I put a println line in every evaluation of WollokInterpreterEvaluator.

Using fqn comparison it works perfect!

```scala
Wollok interactive console (type "quit" to quit):
>>> try { error.throwWithMessage("foo") } catch e : Exception { console.println("jeje") }
jeje
>>> try { error.throwWithMessage("foo") } catch e : Exception { }
null
>>> try { throw new Exception("foo") } catch e : Exception { console.println("jeje") }
jeje
```

**Comment by fdodino** [comment]
Obviously, second example outputs  null, it's not the desired behavior. But we should open a new issue for that.
